Fix cluster alert for watcher/monitoring IndexOutOfBoundsExcep…

jakelandis opened the pull request:
If a cluster sending monitoring data is unhealthy and triggers an
alert, then stops sending data the following exception [1] can occur.

This exception stops the current Watch and the behavior is actually
correct in part due to the exception. [Simply fixing the exception
introduces some incorrect behavior](https://github.com/elastic/elasticsearch/pull/45294). Once the the Watch does not
error in the this case, it will result in an incorrectly "resolved"
alert due to the way alert resolution is coded.  

The fix here is two parts 
a) fix the exception - easy as checking the size of the
array before accessing it

b) fix the incorrect behavior that results from fixing the exception.

- Note - the UI depends on the success/met state for each condition
to determine an "OK" or "FIRING"

In this scenario, where an unhealthy cluster triggers an alert and
then goes silent, it should keep "FIRING" until it hears back that
the cluster is green. To keep the Watch "FIRING" either the index
action or the email action needs to fire. Since the Watch is neither
a "new" alert or a "resolved" alert, we do not want to keep sending
an email (that would be non-passive too). Without completely changing
the logic of how an alert is resolved allowing the index action to
take place would result in the alert being resolved. 

Since we can not keep "FIRING" either the email or index action 
(since we don't want to resolve the alert or re-write the logic for alert resolution),
we will introduce a 3rd action. A logging action that WILL fire when
the cluster is unhealthy. Specifically will fire when there is an
unresolved alert and it can not find the cluster state.
This logging action is logged at debug, so it should not be noticed much.
This logging action serves as an 'anchor' for the UI to keep the state
in an a "FIRING" status until the alert is resolved.

This commit also increases the versions of the watches. 

This presents a possible scenario where a cluster starts firing,
then goes completely silent forever, the Watch will be "FIRING"
forever. This is an edge case that already exists in some scenarios
and requires manual intervention to remove that Watch.

Fixes #43184

[1]
```
org.elasticsearch.script.ScriptException: runtime error
	at org.elasticsearch.painless.PainlessScript.convertToScriptException(PainlessScript.java:94) ~[?:?]
	at org.elasticsearch.painless.PainlessScript$Script.execute(ctx.vars.email_recipient = (ctx.payload.kibana_settings.hits.total > 0 && ctx.payload.kibana_settings.hits.hits[0]._source.kibana_settings.xpack != null) ? ctx.payload.kibana_settings.hits.hits[0]._source.kibana_settings.xpack.default_admin_email : null;ct ...:1152) ~[?:?]
	at org.elasticsearch.xpack.watcher.transform.script.ExecutableScriptTransform.doExecute(ExecutableScriptTransform.java:49) ~[x-pack-watcher-7.1.1.jar:7.1.1]
	...
Caused by: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
```
from .watcher-history:
```
     "type" : "script_exception",
                "reason" : "runtime error",
                "script_stack" : [
                  "java.util.ArrayList.rangeCheck(ArrayList.java:657)",
                  "java.util.ArrayList.get(ArrayList.java:433)",
                  "state = ctx.payload.check.hits.hits[0]._source.cluster_state.status;",
                  "                                   ^---- HERE"
                ],
```

----
To test, enable monitoring and a Kibana email, push the cluster to a unhealthy state:

Kibana
```
xpack.monitoring.cluster_alerts.email_notifications.enabled: true
xpack.monitoring.cluster_alerts.email_notifications.email_address: alert@example.com
```

ES
```
xpack.monitoring.enabled: true
xpack.monitoring.collection.enabled: true
```

Make unhealthy
```
PUT deleteme
{
  "settings": {
    "number_of_replicas": 100
  }
}
```

Stop sending metrics
```
PUT _cluster/settings
{
  "transient": {
    "xpack.monitoring.collection.interval" : "20000s"
  }
}
```
Delete the monitoring index(s)
```
DELETE .monitoring-es-7*
```
Observe that UI stays in a firing state 
[screen shot](https://user-images.githubusercontent.com/976291/62655668-198cd080-b928-11e9-9b5b-f9f6902a85d1.png)

Fix the cluster
```
DELETE deleteme
```
Allowing monitoring data to flow:
```
PUT _cluster/settings
{
  "transient": {
    "xpack.monitoring.collection.interval" : null
  }
}
```
A resolved email should now be sent and the UI is now OK

Transitional phase that sent the resolved
[screen shot](https://user-images.githubusercontent.com/976291/62655673-1e518480-b928-11e9-8712-1ea260a0e731.png)
Stabilized in OK
[screen shot](https://user-images.githubusercontent.com/976291/62655682-214c7500-b928-11e9-96b8-e909135f9ca4.png)

---------
A more readable diff:
<img width="2117" alt="image" src="https://user-images.githubusercontent.com/976291/62656632-3f1ad980-b92a-11e9-81fd-4f3e5160c49b.png">

jakelandis committed:
Fix cluster alert for watcher/monitoring IndexOutOfBoundsException

If a cluster sending monitoring data is unhealthy and triggers an
alert, then stops sending data the following exception [1] can occur.

This exception stops the current Watch and the behavior is actually
correct in part due to the exception. Simply fixing the exception
introduces some incorrect behavior. Now that the Watch does not
error in the this case, it will result in an incorrectly "resolved"
alert.  The fix here is two parts a) fix the exception b) fix the
following incorrect behavior.

a) fixing the exception is as easy as checking the size of the
array before accessing it.

b) fixing the following incorrect behavior is a bit more intrusive

- Note - the UI depends on the success/met state for each condition
to determine an "OK" or "FIRING"

In this scenario, where an unhealthy cluster triggers an alert and
then goes silent, it should keep "FIRING" until it hears back that
the cluster is green. To keep the Watch "FIRING" either the index
action or the email action needs to fire. Since the Watch is neither
a "new" alert or a "resolved" alert, we do not want to keep sending
an email (that would be non-passive too). Without completely changing
the logic of how an alert is resolved allowing the index action to
take place would result in the alert being resolved. Since we can
not keep "FIRING" either the email or index action (since we don't
want to resolve the alert nor re-write the logic for alert resolution),
we will introduce a 3rd action. A logging action that WILL fire when
the cluster is unhealthy. Specifically will fire when there is an
unresolved alert and it can not find the cluster state.
This logging action is logged at debug, so it should be noticed much.
This logging action serves as an 'anchor' for the UI to keep the state
in an a "FIRING" status until the alert is resolved.

This presents a possible scenario where a cluster starts firing,
then goes completely silent forever, the Watch will be "FIRING"
forever. This is an edge case that already exists in some scenarios
and requires manual intervention to remove that Watch.

Fixes #43184

[1]
```
org.elasticsearch.script.ScriptException: runtime error
	at org.elasticsearch.painless.PainlessScript.convertToScriptException(PainlessScript.java:94) ~[?:?]
	at org.elasticsearch.painless.PainlessScript$Script.execute(ctx.vars.email_recipient = (ctx.payload.kibana_settings.hits.total > 0 && ctx.payload.kibana_settings.hits.hits[0]._source.kibana_settings.xpack != null) ? ctx.payload.kibana_settings.hits.hits[0]._source.kibana_settings.xpack.default_admin_email : null;ct ...:1152) ~[?:?]
	at org.elasticsearch.xpack.watcher.transform.script.ExecutableScriptTransform.doExecute(ExecutableScriptTransform.java:49) ~[x-pack-watcher-7.1.1.jar:7.1.1]
	...
Caused by: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
```
from .watcher-history:
```
     "type" : "script_exception",
                "reason" : "runtime error",
                "script_stack" : [
                  "java.util.ArrayList.rangeCheck(ArrayList.java:657)",
                  "java.util.ArrayList.get(ArrayList.java:433)",
                  "state = ctx.payload.check.hits.hits[0]._source.cluster_state.status;",
                  "                                   ^---- HERE"
                ],
```

jakelandis committed:
bump to 8.0.0

elasticmachine commented:
Pinging @elastic/es-core-features

ycombinator commented on x-pack/plugin/monitoring/src/main/resources/monitoring/watches/elasticsearch_cluster_status.json:
```diff
@@ -134,19 +134,31 @@
   "actions": {
+    "log_state_not_found": {
+      "condition": {
+        "script": "!ctx.vars.found_state"
```
Shouldn't this condition be `ctx.vars.found_state == "unknown"`?

ycombinator commented on x-pack/plugin/monitoring/src/main/resources/monitoring/watches/elasticsearch_cluster_status.json:
```diff
@@ -134,19 +134,31 @@
+        "script": "!ctx.vars.found_state"
+      },
+      "logging" : {
+        "text" : "Watcher could not determine cluster state for cluster [{{ctx.metadata.xpack.cluster_uuid}}]. This likely means the cluster has not sent any monitoring data recently.",
```
WDYT about adding `{{ctx.metadata.xpack.watch}}` in here somewhere, just so we can tell this log message is coming from a Stack Monitoring Cluster Alert?

jakelandis replied:
done on https://github.com/elastic/elasticsearch/pull/45308/commits/73635f3d5766046823f359cd5d5e9a1cb1cad973

jakelandis committed:
update add {{ctx.metadata.xpack.watch}} to log message

ycombinator reviewed (approved):
I can confirm that the `org.elasticsearch.script.ScriptException: runtime error` has gone away with this PR and the UI behavior remains unchanged.

Code changes LGTM too.

When backporting this PR I assume `LAST_UPDATED_VERSION` will need to be adjusted?

jakelandis commented:
thanks @ycombinator - yes, both the `LAST_UPDATED_VERSION` and version inside the watch itself will be updated accordingly. (I infer that is the expected design to prevent running newer watches (from remote clusters) then the current node that is running them).

rjernst commented on x-pack/plugin/monitoring/src/main/resources/monitoring/watches/elasticsearch_nodes.json:
```diff
@@ -7,7 +7,7 @@
       "severity": 1999,
       "type": "monitoring",
-      "version_created": 7000099,
+      "version_created": 8000099,
```
Could this be replaced when the watch is loaded, similar to watch id? Being hardcoded seems to make this prone to getting out of date.

jakelandis replied:
++ , however, I would prefer to address that on a separate PR.

rjernst replied:
Is this version correct though? Won't this PR be backported to 7.x?

jakelandis replied:
The version is correct for master. I will have to manually change this in the backport. FWIW, i will need to manually change it with the templated approach too, just in 1 place rather then 6. 

This version is used to determine if the watches need to be updated, so we don't want to set to Version.Current otherwise every version update would update them. (which is i guess isn't that bad)

rjernst replied:
If the intention is to have the version created match when the template was changed, wouldn't that be 7.5.0 (assuming backporting to 7.x right now)? Having this version be different in master and 7.x even though the template is the same seems to defeat the purpose?

jakelandis replied:
thanks, I see your point now.  I will close out  #47744 and incorporate it here with the version set to 7.5.0

jakelandis replied:
Changes merged in and found at https://github.com/elastic/elasticsearch/pull/45308/commits/9d66e2e5e97dab8c314a163c45f29dbcb529d4f2 and https://github.com/elastic/elasticsearch/pull/45308/commits/fcb5c4086acf243e67d1a16161e1239d751c24c6

jakelandis committed:
Merge branch 'master' into monitoring_watcher_cluster_alert_out_of_bounds2

jakelandis commented on x-pack/plugin/monitoring/src/main/resources/monitoring/watches/elasticsearch_cluster_status.json:
```diff
@@ -134,19 +134,31 @@
   "transform": {
     "script": {
-      "source": "ctx.vars.email_recipient = (ctx.payload.kibana_settings.hits.total > 0 && ctx.payload.kibana_settings.hits.hits[0]._source.kibana_settings.xpack != null) ? ctx.payload.kibana_settings.hits.hits[0]._source.kibana_settings.xpack.default_admin_email : null;ctx.vars.is_new = ctx.vars.fails_check && !ctx.vars.not_resolved;ctx.vars.is_resolved = !ctx.vars.fails_check && ctx.vars.not_re ...
+      "source": "ctx.vars.email_recipient = (ctx.payload.kibana_settings.hits.total > 0 && ctx.payload.kibana_settings.hits.hits[0]._source.kibana_settings.xpack != null) ? ctx.payload.kibana_settings.hits.hits[0]._source.kibana_settings.xpack.default_admin_email : null;ctx.vars.is_new = ctx.vars.fails_check && !ctx.vars.not_resolved;ctx.vars.is_resolved = !ctx.vars.fails_check && ctx.vars.not_re ...
```
> if (ctx.vars.not_resolved){ctx.payload = ctx.payload.alert.hits.hits[0]._source ... Is this still not guarded for no hits existing?

This is NPE safe since `ctx.vars.not_resolved` is defined in the condition as:
```
ctx.payload.alert.hits.total == 1 && ...
```

Therefor the `ctx.payload.alert.hits.hits` is always non-null if wrapped in `if(ctx.vars.not_resolved)`

jakelandis commented:
@rjernst - Thanks for your review, and apologies for letting this slip off my radar for so long. I addressed the two comments and plan to submit a separate PR to address the hard coded version number. Would you mind taking another look ?

jakelandis committed:
Remove hard coded version_created in default monitoring alerts

This change uses a template-like method to populate the version_created
for the default monitoring watches. This value is used to know if the
watches should be updated. Keeping this current to at least the major
version helps to ensure that the documents that store the watches are
updated at least once per major version.

jakelandis commented:
PR to fix the hard coded values in the watch templates. https://github.com/elastic/elasticsearch/pull/47744

jakelandis committed:
change version changed to 7.5.0

rjernst reviewed (approved):
LGTM. Long term I wonder if we should be using the version id there or if we could use the string version (eg 7.5.0). The id is really an internal implementation detail, and the more places it exists the more difficult it is to change. The 99 suffix is no longer needed, but removing it is going to be quite difficult. Just a thought for future work around here.
